docs/reference/content/bson/macros.md

```diff
@@ -0,0 +1,86 @@
++++
+date = "2016-11-20T14:14:00-00:00"
+title = "Macros"
+[menu.main]
+  parent = "BSON"
+  weight = 30
+  pre = "<i class='fa fa-cog'></i>"
++++
+
+## Macros
+
+New in 2.0 the scala driver allows you to use case classes to represent documents in a collection via the 
+[`Macros`]({{< apiref "org.mongodb.scala.bson.codecs.Macros" >}}) helper.  Simple case classes and nested case classes are supported. 
+Hierarchical modelling can be achieve by using a sealed trait and having case classes implement the parent trait.
+
+Many simple scala types are supported and they will be marshaled into their corresponding 
+[`BsonValue`]({{< apiref "org.mongodb.scala.bson.BsonValue" >}}) type. Below is a list of Scala types and their type safe Bson representation:
+   
+| Scala type                        | BSON type         |
+|-----------------------------------|-------------------|
+| case class                        | Document          |
+| `Iterable`                        | Array             |
+| `Date`                            | Date              |
+| `Boolean`                         | Boolean           |
+| `Double`                          | Double            |
```

Comment:
We have a small problem with the numeric types.  Consider a case class like:

    case class Numeric(d : Int)

and a document that was created in the shell like:

    db.test.insert({ d : 1})

Since the shell will encode the value as a double, upon decoding we'll get an exception:

    org.bson.BsonInvalidOperationException: 
        readInt32 can only be called when CurrentBSONType is INT32, 
        not when CurrentBSONType is DOUBLE

To fix problems of this ilk, this, we will have to modify the numeric Codec implementations in the Java driver to handle all the numeric types on decode.  See the [C# driver Int32Serializer](https://github.com/mongodb/mongo-csharp-driver/blob/ee97df791d0a42853295bcc5cd791e37b8bb076d/src/MongoDB.Bson/Serialization/Serializers/Int32Serializer.cs#L106-L131) for one way of doing this.